chain-abci/src/app/app_init.rs

```diff
@@ -443,119 +467,117 @@ impl<T: EnclaveProxy> ChainNodeApp<T> {
-
-                tendermint_validator_addresses
-                    .insert(tendermint_validator_address.clone(), *address);
-
-                validator_liveness.insert(
-                    tendermint_validator_address,
-                    LivenessTracker::new(network_params.get_block_signing_window()),
-                );
-            }
+        let network_params = NetworkParameters::Genesis(conf.network_params);
+        let new_account_root = compute_accounts_root(&mut self.accounts, &accounts);
+        let genesis_app_hash = compute_app_hash(
+            &MerkleTree::empty(),
+            &new_account_root,
+            &rp,
+            &network_params,
+        );
 
-            // check req.validators is consistent with app_state's council nodes
-            let mut req_validators = req.validators.clone().into_vec();
-            let fn_sort_key = |a: &ValidatorUpdate| {
-                a.pub_key
-                    .as_ref()
-                    .map(|key| (key.field_type.clone(), key.data.clone()))
-            };
-            validators.sort_by_key(fn_sort_key);
-            req_validators.sort_by_key(fn_sort_key);
-            if validators != req_validators {
-                panic!("validators in genesis configuration are not consistent with app_state");
-            }
+        if self.genesis_app_hash != genesis_app_hash {
+            panic!("initchain resulting genesis app hash: {} does not match the expected genesis app hash: {}", hex::encode(genesis_app_hash), hex::encode(self.genesis_app_hash));
+        }
 
-            let genesis_state = ChainNodeState::genesis(
-                genesis_app_hash,
-                genesis_time,
-                new_account_root,
-                rp,
-                network_params,
-                ValidatorState {
-                    council_nodes_by_power: validator_by_voting_power,
-                    tendermint_validator_addresses,
-                    punishment: ValidatorPunishment {
-                        validator_liveness,
-                        slashing_schedule: Default::default(),
-                    },
-                },
+        let mut inittx = db.transaction();
+        check_and_store_consensus_params(
+            req.consensus_params.as_ref(),
+            &nodes,
+            &network_params,
+            &mut inittx,
+        );
+
+        let mut validators = Vec::with_capacity(nodes.len());
+        let mut validator_liveness = BTreeMap::new();
+        let mut validator_by_voting_power = BTreeMap::new();
+        let mut tendermint_validator_addresses = BTreeMap::new();
+
+        for (address, node) in nodes.iter() {
+            let mut validator = ValidatorUpdate::default();
+            let power = get_voting_power(&conf.distribution, address);
+            self.validator_voting_power.insert(*address, power);
+            validator_by_voting_power.insert((power, *address), node.clone());
+            validator.set_power(power.into());
```

Comment:
why is it creating validator updates here when at the end, it does `ResponseInitChain::new()` which won't contain this?

Reply:
It's only for validating `req.validators` is correct.